src/mono/mono/metadata/object.c

```diff
@@ -5652,7 +5653,8 @@ mono_runtime_try_invoke_array (MonoMethod *method, void *obj, MonoArray *params,
 		for (i = 0; i < mono_array_length_internal (params); i++) {
 			MonoType *t = sig->params [i];
 			pa [i] = invoke_array_extract_argument (params, i, t, &has_byref_nullables, error);
-			return_val_if_nok (error, NULL);
+			MONO_HANDLE_PIN (pa [i]);
```

Comment:
pa[i] might not be an object reference.

Reply:
This is "fine" because we conservatively scan the handles. Would it be better to have the callee (invoke_array_extract_argument) explicitly create a new handle for every case where it returns an object ? This follows the above question about who guarantees the liveness of the returned value.

Reply:
I'd suggest returning the ref as an out argument from invoke_array_extract_argument():
```
MonoObject *pa_obj = NULL;
pa [i] = invoke_array_extract_argument (params, i, t, &pa_obj, &has_byref_nullables, error);
if (pa_obj)
  MONO_PIN (pa_obj);
```